src/contracts/socket/index.ts

```diff
@@ -0,0 +1,166 @@
+import { SupportedChainId } from '@cowprotocol/cow-sdk';
+import {
+  Contract as WeirollContract,
+  Planner as WeirollPlanner,
+} from '@weiroll/weiroll.js';
+import { ethers } from 'ethers';
+import { BaseTransaction } from '../../types';
+import { getWallet } from '../../utils';
+import { getCowShedAccount } from '../cowShed';
+import { getErc20Contract } from '../erc20';
+import { CommandFlags, getWeirollTx } from '../weiroll';
+import { bytesLibAbi, socketGatewayAbi } from './types';
+import {
+  decodeBungeeTxData,
+  getBungeeQuote,
+  getBungeeRouteTransactionData,
+} from './utils';
+
+export interface BridgeWithBungeeParams {
+  owner: string;
+  sourceChain: SupportedChainId;
+  sourceToken: string;
+  sourceTokenAmount: bigint;
+  targetToken: string;
+  targetChain: number;
+  recipient: string;
+  includeBridges: string[];
+}
+
+export async function bridgeWithBungee(
+  params: BridgeWithBungeeParams
+): Promise<BaseTransaction> {
+  const {
+    owner,
+    sourceChain,
+    sourceToken,
+    sourceTokenAmount,
+    targetChain,
+    targetToken,
+    recipient,
+    includeBridges,
+  } = params;
+
+  // Get cow-shed account
+  const cowShedAccount = getCowShedAccount(sourceChain, owner);
+
+  const planner = new WeirollPlanner();
+
+  // Get bungee quote
+  const quote = await getBungeeQuote({
+    fromChainId: sourceChain.toString(),
+    fromTokenAddress: sourceToken,
+    toChainId: targetChain.toString(),
+    toTokenAddress: targetToken,
+    fromAmount: sourceTokenAmount.toString(),
+    userAddress: cowShedAccount, // bridge input token will be in cowshed account
+    recipient: recipient,
+    sort: 'output', // optimize for output amount
+    singleTxOnly: true, // should be only single txn on src chain, no destination chain txn
+    isContractCall: true, // get quotes that are compatible with contracts
+    disableSwapping: true, // should not show routes that require swapping
+    includeBridges,
+  });
+  if (!quote) {
+    throw new Error('No quote found');
+  }
+  console.log('🔗 Socket quote:', quote);
+  // check if routes are found
+  if (!quote.result.routes.length) {
+    throw new Error('No routes found');
+  }
+  // check if only single user tx is present
+  if (quote.result.routes[0].userTxs.length > 1) {
+    throw new Error('Multiple user txs found');
+  }
+  // check if the user tx is fund-movr
+  if (quote.result.routes[0].userTxs[0].userTxType !== 'fund-movr') {
+    throw new Error('User tx is not fund-movr');
+  }
+
+  // use the first route to prepare the bridge tx
+  const route = quote.result.routes[0];
+  const txData = await getBungeeRouteTransactionData(route);
+  const { routeId, encodedFunctionData } = decodeBungeeTxData(
+    txData.result.txData
+  );
```

Comment:
Will we be able to verify this data to guarantee we don't ask users to blind-sign on something returned from the API?

Reply:
We've added a calldata verification step in our PoC script

Basically we have a verifier contract which you could input the encoded calldata received from our backend and the expected values for key params like the input amount, recipient, destination chain id, token contract etc. and the verifier will validate if the encoded calldata does match the expected values. 

You'll find it implemented in this commit: b16b3ff1386c2793c49b283215b273afba3fdd92

These are the relevant verified contracts:
SocketVerifier: https://arbiscan.io/address/0x69D9f76e4cbE81044FE16C399387b12e4DBF27B1#code
AcrossV3Verification: https://arbiscan.io/address/0x2493Ac43A301d0217abAD6Ff320C5234aEe0931d#code
CCTPVerification: https://arbiscan.io/address/0xF58Db19f264359D6687b5693ee516bf316BE3Ba6#code

Reply:
Let us know if this helps the verifiability of the bridge txn calldata

Reply:
That's great, thank you!

BTW, there's a typo in this contract function: `validateRotueId`
Better to fix it while it still in a single chain IMO :)

Reply:
Actually `SocketVerifier` is already deployed on a few chains like arbitrum, optimism, polygon at the same address for a few other bridges and actively used for similar purposes

Reply:
Oh I see! I guess I'll have to fight with my OCD and accept `rotue` then 😅